src/main/java/seedu/address/logic/commands/EditCommand.java

```diff
@@ -109,8 +109,14 @@ public CommandResult execute(Model model) throws CommandException {
             Person personToEdit = lastShownList.get(index.getZeroBased());
             Person editedPerson = createEditedPerson(personToEdit, editPersonDescriptor);
 
-            if (!personToEdit.isSamePerson(editedPerson) && model.hasPerson(editedPerson)) {
-                throw new CommandException(MESSAGE_DUPLICATE_PERSON);
+            if (!personToEdit.equals(editedPerson)) {
+                if (model.getAddressBook()
+                        .getPersonList()
+                        .stream()
+                        .filter(person -> person != personToEdit)
+                        .anyMatch(editedPerson::isSamePerson)) {
+                    throw new CommandException(MESSAGE_DUPLICATE_PERSON);
+                }
```

Comment:
Just to explain what happened:

In short, when should a `MESSAGE_DUPLICATE_PERSON` need to be thrown? In our app, only two command could possible caused it to throw: `add` and `edit`.

For `add`, a newly added person cannot be same as any existing person, so the code
```java
if (model.hasPerson(toAdd)) {
    throw new CommandException(MESSAGE_DUPLICATE_PERSON);
}
```
works nicely.

For `edit`, a duplicate happens only if a person changes phone, email, or telegram, such that one of these fields are same as **other** existing person. Noted that we have two things to check:
1. A person changes phone, email, or telegram.
2. One of these fields are same as **other** existing person.

To achieve point 1, the AB3 code `!personToEdit.isSamePerson(editedPerson)`only suggests that a person is no longer same if and only if all three fields (phone, email, telegram) are changed. So if any of these three fields are changed (but not all three), it will not be able to catch it. Hence, we use `!personToEdit.equals(editedPerson)`, so that any change can be capture as long as the person is edited.

To achieve point 2, imagine the original person list that contains all other existing person, and the original version of the person being edited. The AB3 code `model.hasPerson(editedPerson)` checks whether the editedPerson is in the original list. Since the AB3 code only defines two person that shares same name to be same, and we have checked that the person changes name at the first part, of course `model.hasPerson(editedPerson)` will "bypass" the original version of the person being edited. 

But in our case, due to the implementation of `isSamePerson()`, we should explicitly exclude the original version of the person being edited from the list to avoid problem.